File: src/jastudio_src/jastudio/language_services/janome_ex/tokenizing/jn_tokenized_text.py
from __future__ import annotations

from typing import TYPE_CHECKING

from autoslot import Slots  # pyright: ignore[reportMissingTypeStubs]
from jastudio import mylog
from jastudio.ankiutils import app
from jastudio.language_services.janome_ex.tokenizing.pre_processing_stage.pre_processing_stage import PreProcessingStage

if TYPE_CHECKING:
    from janome.tokenizer import Token  # pyright: ignore[reportMissingTypeStubs]
    from jastudio.language_services.janome_ex.tokenizing.analysis_token import IAnalysisToken
    from jastudio.language_services.janome_ex.tokenizing.jn_token import JNToken

class JNTokenizedText(Slots):
    def __init__(self, text: str, raw_tokens: list[Token], tokens: list[JNToken]) -> None:
        self.raw_tokens: list[Token] = raw_tokens
        self.text: str = text
        self.tokens: list[JNToken] = tokens

    def pre_process(self) -> list[IAnalysisToken]:
        try:
            return PreProcessingStage(app.col().vocab).pre_process(self.tokens)
        except Exception:
            mylog.error(f"""Failed to pre-process text: {self.text}""")
            raise

File: src/src_dotnet/JAStudio.Core/LanguageServices/JanomeEx/Tokenizing/JNTokenizedText.cs
using System;
using System.Collections.Generic;
using JAStudio.Core.LanguageServices.JamdictEx;
using JAStudio.Core.Note.Collection;

namespace JAStudio.Core.LanguageServices.JanomeEx.Tokenizing;

public class JNTokenizedText
{
    public string Text { get; }
    public List<JNToken> Tokens { get; }
    public object RawTokens { get; } // Python-specific Janome tokens, kept as object for compatibility

    public JNTokenizedText(string text, List<JNToken> tokens, object? rawTokens = null)
    {
        Text = text;
        Tokens = tokens;
        RawTokens = rawTokens ?? new object(); // Placeholder - not used in C# but maintained for API compatibility
    }

    public List<IAnalysisToken> PreProcess(VocabCollection vocab, DictLookup dictLookup)
    {
        try
        {
            return new PreProcessingStage.PreProcessingStage(vocab, dictLookup).PreProcess(Tokens);
        }
        catch (Exception)
        {
            MyLog.Error($"Failed to pre-process text: {Text}");
            throw;
        }
    }
}
